src/compiler/csharp_generator.cc

```diff
@@ -323,8 +323,52 @@ std::vector<const Descriptor*> GetUsedMessages(
   return result;
 }
 
-void GenerateMarshallerFields(Printer* out, const ServiceDescriptor* service) {
+void GenerateMarshallerFields(Printer* out, const ServiceDescriptor* service,
+                              bool use_buffer_serialization) {
   std::vector<const Descriptor*> used_messages = GetUsedMessages(service);
+  out->Print("#if !GOOGLE_PROTOBUF_DISABLE_BUFFER_SERIALIZATION\n");
+  // Generate buffer serialization marshallers
+  for (size_t i = 0; i < used_messages.size(); i++) {
+    const Descriptor* message = used_messages[i];
+    out->Print(
+        "static readonly grpc::Marshaller<$type$> $fieldname$ = \n",
+        "fieldname", GetMarshallerFieldName(message), "type",
+        GetClassName(message));
+    out->Indent();
+    out->Print(
+          "grpc::Marshallers.Create(\n");
+    out->Indent();
+    out->Print(
+            "(arg, context) =>\n"
+            "{\n");
+    out->Indent();
+    out->Print(
+              "var writer = new global::Google.Protobuf.CodedOutputWriter("
+              "context.GetBufferWriter());\n"
+              "arg.WriteTo(ref writer);\n"
+              "writer.Flush();\n"
+              "context.Complete();\n");
+    out->Outdent();
+    out->Print(
+            "},\n"
+            "context =>\n"
+            "{\n");
+    out->Indent();
+    out->Print(
+              "var result = new $type$();\n"
```

Comment:
Use `IMessageParser<T>.ParseFrom(ReadOnlySequence<byte>)` here. e.g.

```cs
(context) => global::Math.DivArgs.Parser.ParseFrom(context.PayloadAsReadOnlySequence())
```

The helper method will create the reader for you.

Reply:
By the way, this method will try to use IBufferMessage, and fallback to IMessage.